### PR TITLE
Element: add sdf::Errors output to API methods

### PR DESCRIPTION
**Pull request by marcoag**
Signed-off-by: Marco A. Gutierrez <marco@openrobotics.org>

# 🎉 New feature

Work towards #820

## Summary
Adds `sdf::Errors` for as an error report option to all the methods on the Element class.

## Test it
Calling the new functions that contain the `sdf::Errors` parameter should only add errors to the `sdf::Errors` data structure and not produce any printed output or throw any exceptions.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: include/sdf/Error.hh [hunk 1]
```diff
@@ -85,6 +85,9 @@ namespace sdf
     /// expected, and an element of a different type was received.
     ELEMENT_INCORRECT_TYPE,
 
+    /// \Briev Generic error type for Elements
```
Comment (scpeters) [inline include/sdf/Error.hh:88]:
```suggestion
    /// \brief Generic error type for Elements
```

File: include/sdf/Element.hh [hunk 2]
```diff
@@ -83,10 +83,20 @@ namespace sdf
...
+    public: ElementPtr Clone(sdf::Errors &_errors) const;
+
     /// \brief Copy values from an Element.
     /// \param[in] _elem Element to copy value from.
     public: void Copy(const ElementPtr _elem);
 
+    /// \brief Copy values from an Element.
+    /// \param[in] _elem Element to copy value from.
+    /// \param[out] _errors Vector of errors.
+    public: void Copy(const ElementPtr _elem, sdf::Errors &_errors);
```
Comment (scpeters) [inline include/sdf/Element.hh:98]: Since this is targeting `main`, we can change API if we do it this week. For any of these functions that currently return `void`, we could change the return type to `sdf::Errors` to avoid duplicating the functions

Comment (marcoag) [reply include/sdf/Element.hh:98]: The idea was to still maintain the old output as an option. If we have `sdf::Errors` as a return and only one function then I believe we would need a way to select one of the two ways of error report. I can think of a boolean flag but doesn't sound very appealing, any other way to solve this?

File: include/sdf/Error.hh [hunk 3]
```diff
@@ -85,6 +85,9 @@ namespace sdf
     /// expected, and an element of a different type was received.
     ELEMENT_INCORRECT_TYPE,
 
+    /// \brief Generic error type for Elements
+    ELEMENT_ERROR,
```
Comment (scpeters) [inline include/sdf/Error.hh:89]: actually inserting an enum value breaks API; I'll split this into a separate PR

Comment (scpeters) [reply include/sdf/Error.hh:89]: #1099

File: src/Element.cc [hunk 4]
```diff
@@ -170,6 +201,25 @@ ParamPtr Element::CreateParam(const std::string &_key,
...
 
+/////////////////////////////////////////////////
+ParamPtr Element::CreateParam(const std::string &_key,
+                              const std::string &_type,
+                              const std::string &_defaultValue,
+                              bool _required,
+                              sdf::Errors &_errors,
+                              const std::string &_description)
+{
+  ParamPtr param = std::make_shared<Param>(
```
Comment (azeey) [inline src/Element.cc:212]: This is duplicating code. See my comments on `AddValue`

File: src/Element.cc [hunk 5]
```diff
@@ -181,8 +231,33 @@ void Element::AddAttribute(const std::string &_key,
...
 
+/////////////////////////////////////////////////
+void Element::AddAttribute(const std::string &_key,
+                           const std::string &_type,
+                           const std::string &_defaultValue,
+                           bool _required,
+                           sdf::Errors &_errors,
+                           const std::string &_description)
+{
+  this->dataPtr->attributes.push_back(
```
Comment (azeey) [inline src/Element.cc:242]: This is duplicating code. See my comments on `AddValue`.

File: src/Element.cc [hunk 6]
```diff
@@ -941,6 +1109,18 @@ ElementPtr Element::GetElement(const std::string &_name)
   return result;
 }
 
+/////////////////////////////////////////////////
+ElementPtr Element::GetElement(const std::string &_name, sdf::Errors &_errors)
+{
+  ElementPtr result = this->GetElementImpl(_name);
```
Comment (azeey) [inline src/Element.cc:1115]: This is duplicating code. See my comments on `AddValue`.

File: src/gz_TEST.cc [hunk 7]
```diff
@@ -1871,6 +1875,8 @@ TEST(inertial_stats, GZ_UTILS_TEST_DISABLED_ON_WIN32(SDF))
     std::string output =
       custom_exec_str(GzCommand() + " sdf --inertial-stats " +
                       path + SdfVersion());
+    std::cout << GzCommand() + " sdf --inertial-stats " +
```
Comment (azeey) [inline src/gz_TEST.cc:1878]: nit: Remove debug code?

File: test/integration/error_output.cc [hunk 8]
```diff
@@ -106,7 +106,8 @@ TEST(Error, ErrorOutput)
   // will make reparse fail
   sdf::Param poseParam("", "pose", "1 2 3 0.4 0.5 0.6", false, "description");
   sdf::ElementPtr poseElem(new sdf::Element);
-  poseElem->AddAttribute("rotation_format", "string", "invalid_format", false);
+  poseElem->AddAttribute(
```
Comment (azeey) [inline test/integration/error_output.cc:109]: nit: Is this change necessary?

**Review by azeey — CHANGES_REQUESTED** [review]
Looks good! I have some comments about deduplicating code and avoiding extra new member functions.

File: include/sdf/Element.hh [hunk 9]
```diff
@@ -610,6 +698,51 @@ namespace sdf
...
+    /// \brief Create a new Param object and return it.
+    /// \param[in] _key Key for the parameter.
+    /// \param[in] _type String name for the value type (double,
+    /// int,...).
+    /// \param[in] _defaultValue Default value.
+    /// \param[in] _required True if the parameter is required to be set.
+    /// \param[out] _errors Vector of errors.
+    /// \param[in] _description Description of the parameter.
+    /// \return A pointer to the new Param object.
+    private: ParamPtr CreateParam(const std::string &_key,
```
[the remark this answers is not in the document]
Comment (marcoag) [reply include/sdf/Element.hh:710]: This one seems to use `SDF_ASSERT`, which throws an `sdf::AssertionInternalError`, changing the function to use `sdf::Error` instead will change behavior of any other function using `CreateParam`... dunno if we are ok with that, are we?

File: src/Element.cc [hunk 10]
```diff
@@ -141,6 +141,17 @@ void Element::AddValue(const std::string &_type,
...
 }
 
+/////////////////////////////////////////////////
+void Element::AddValue(const std::string &_type,
+                       const std::string &_defaultValue,
+                       bool _required,
+                       sdf::Errors &_errors,
+                       const std::string &_description)
+{
+  this->dataPtr->value = this->CreateParam(this->dataPtr->name,
```
[the remark this answers is not in the document]
Comment (marcoag) [reply src/Element.cc:151]: Just realized that errors are sometimes using `SDF_ASSERT`  and others the `sdferr` stream. The first one throws an exception while the later prints using `sdf::Console`. Since these are two different behaviors I guess we can do either one of these:
* we don't care about changing this behavior and use either `sdferr` or `SDF_ASSERT` by default in the calling function.
* we add some "extra" info to `sdf::Errors`, maybe by creating a new enum that says `ASSERT_ERROR` or `SDFERR_...` so we can differentiate between both of them and keep the same behavior.

I also realized this might change behavior since a call to `SDF_ASSERT` will break the normal execution flow whereas if we add errors to `sdf::Errors` the function will keep executing. I think this can be solved with just a `return` and then with the throwing the `SDF_ASSERT` in the calling function but I just wanted to leave it here in case I miss something.

File: include/sdf/Element.hh (hunk 9 above)
[the remark this answers is not in the document]
Comment (azeey) [reply include/sdf/Element.hh:710]: That's true. If Garden wasn't feature frozen, I'd say we should remove the SDF_ASSERT, but now that this is going in a minor release, I don't think we should change behavior like that. So, yeah, let's keep the second `CreateParam`, but I'd still think we should avoid duplicating code and write one `CreateParam` by calling the other.

File: src/Element.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (azeey) [reply src/Element.cc:151]: Like I mentioned in another comment, I'd say let's keep the SDF_ASSERT. We can still call `AddValue` with a throwaway `Errors` object and call `SDF_ASSERT` if there are any errors.

**Comment by chapulina** [comment]
Removed the `Garden` label because we're past feature freeze ❄️ .

This PR can either be merged into `main` and backported to `sdf13` after the stable release, or retargeted to `sdf13` and merged after the stable release (i.e. October).

File: src/Element.cc [hunk 11]
```diff
@@ -975,6 +1155,30 @@ bool Element::HasElementDescription(const std::string &_name) const
 
 /////////////////////////////////////////////////
 ElementPtr Element::AddElement(const std::string &_name)
+{
+  ElementPtr elem = this->AddElementImpl(_name);
+  if (elem.get() == nullptr)
+  {
+    sdferr << "Missing element description for [" << _name << "]\n";
+  }
+  return elem;
```
[the remark this answers is not in the document]
Comment (marcoag) [reply src/Element.cc:1164]: Yes, using the other other `AddElement` instead.

File: test/integration/error_output.cc (hunk 8 above)
Comment (marcoag) [reply test/integration/error_output.cc:109]: no, removed.

File: include/sdf/Element.hh [hunk 12]
```diff
@@ -610,6 +698,51 @@ namespace sdf
...
+    /// \brief Create a copy of this Element.
+    /// \param[out] _errors Vector of errors.
+    /// \return A copy of this Element, NULL if there was an error.
+    private: ElementPtr CloneImpl(sdf::Errors *const _errors = nullptr) const;
+
+    /// \brief Copy values from an Element.
+    /// \param[in] _elem Element to copy value from.
+    /// \param[out] _errors Vector of errors.
+    private: void CopyImpl(const ElementPtr _elem,
+                sdf::Errors *const _errors = nullptr);
```
[the remark this answers is not in the document]
Comment (marcoag) [reply include/sdf/Element.hh:745]: Yes, I'll do that.

File: src/gz_TEST.cc (hunk 7 above)
Comment (marcoag) [reply src/gz_TEST.cc:1878]: Removed.

File: src/Element.cc (hunk 6 above)
Comment (marcoag) [reply src/Element.cc:1115]: Removed duplicating code.

File: src/Element.cc (hunk 5 above)
Comment (marcoag) [reply src/Element.cc:242]: Yes, fixed.

File: src/Element.cc (hunk 4 above)
Comment (marcoag) [reply src/Element.cc:212]: Since this is private, I'll go ahead and change the function to just use `sdf::Errors`.

File: include/sdf/Element.hh (hunk 9 above)
[the remark this answers is not in the document]
Comment (marcoag) [reply include/sdf/Element.hh:710]: Ok then. I went and created an `sdf::ErrorCode::THROW_ERROR` so the calling function can identify which errors to throw. Also added a `throwOrPrintErrors` function to `Utils.cc` to handle the printing or throwing as needed.

File: src/Element.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (marcoag) [reply src/Element.cc:151]: Yes, will try to make sure we keep the same behavior.

File: src/Element.cc [hunk 13]
```diff
@@ -1191,7 +1303,20 @@ void Element::RemoveFromParent()
 /////////////////////////////////////////////////
 void Element::RemoveChild(ElementPtr _child)
 {
-  SDF_ASSERT(_child, "Cannot remove a nullptr child pointer");
+  sdf::Errors errors;
+  RemoveChild(_child, errors);
```
Comment (azeey) [inline src/Element.cc:1307]: Missing `throwOrPrintErrors`

File: src/Utils.hh [hunk 14]
```diff
@@ -85,6 +85,11 @@ namespace sdf
     const sdf::Error &_error,
     sdf::Errors &_errors);
 
+  /// \brief It will print the errors to sdferr or throw them using
+  /// SDF_ASSERT depeneing on their ErrorCode.
```
Comment (azeey) [inline src/Utils.hh:89]: nit:
```suggestion
  /// SDF_ASSERT depending on their ErrorCode.
```

File: src/Utils.hh [hunk 15]
```diff
@@ -85,6 +85,11 @@ namespace sdf
...
+  /// \param[in] _errors  The vector of errors.
+  void throwOrPrintErrors(const sdf::Errors _errors);
```
Comment (azeey) [inline src/Utils.hh:91]: nit: use const ref
```suggestion
  void throwOrPrintErrors(const sdf::Errors& _errors);
```

File: test/integration/error_output.cc [hunk 16]
```diff
@@ -146,3 +146,76 @@ TEST(Error, ErrorOutput)
...
+
+  sdf::Errors errors;
+  sdf::ElementPtr elem = std::make_shared<sdf::Element>();
+  elem->SetName("testElement");
+
+  elem->GetAny(errors, "test");
+  EXPECT_EQ(errors[0].Code(), sdf::ErrorCode::ELEMENT_ERROR);
+  EXPECT_NE(std::string::npos, errors[0].Message().find(
+    "Unable to find value for key [test]"));
+  ASSERT_EQ(errors.size(), 1u);
```
Comment (azeey) [inline test/integration/error_output.cc:166]: The `ASSERT` calls that check the size of the `errors` should come before indexing into it. Otherwise, we might get a segfault.

**Review by azeey — CHANGES_REQUESTED** [review]

File: include/sdf/Error.hh [hunk 17]
```diff
@@ -161,6 +161,10 @@ namespace sdf
     /// \brief The specified parameter (values of SDFormat elements
     /// or attributes) type is unknown.
     UNKNOWN_PARAMETER_TYPE,
+
+    /// \brief generic error to be thrown with SDF_ASSERT by the caller.
+    /// This has been created to help preserve behavior.
+    THROW_ERROR,
```
[the remark this answers is not in the document]
Comment (marcoag) [reply include/sdf/Error.hh:167]: Ok, sounds good.

File: src/Element.cc (hunk 13 above)
Comment (marcoag) [reply src/Element.cc:1307]: Added.

File: test/integration/error_output.cc (hunk 16 above)
Comment (marcoag) [reply test/integration/error_output.cc:166]: True, my bad :man_facepalming:.

**Review by azeey — APPROVED** [review]
Great work!

**Comment by azeey** [comment]
Can you fix the DCO failure? I think it's due to applied suggestions :(

**Comment by marcoag** [comment]
Oh the suggestions DCO thing, I need to be more careful with that :man_facepalming:.

Fixed two out of the three DCOs failures. The last one, commit sha: [96286da](https://github.com/gazebosim/sdformat/pull/1095/commits/96286da3e45156ca9c1a63438d6f84a9ad66c3b7), gets tricky as there are other commits on top of it that I did not authored (see [history](https://github.com/gazebosim/sdformat/commits/marcoag/sdf_error_element)) so rebasing is not an option. The only solutions to this problem I know are either taking it as it is or creating a new PR with the cherry picked relevant commits, any preference?

**Comment by chapulina** [comment]
> The only solutions to this problem I know are either taking it as it is or creating a new PR with the cherry picked relevant commits, any preference?

Another option is adding the other author's signature to the commit as you rebase. I often do that.

**Comment by scpeters** [comment]
> > The only solutions to this problem I know are either taking it as it is or creating a new PR with the cherry picked relevant commits, any preference?
> 
> Another option is adding the other author's signature to the commit as you rebase. I often do that.

you have my permission to add the signature or to just squash it because it is a trivial change. I think rebasing would be a reasonable step

**Comment by scpeters** [comment]
> Can you fix the DCO failure? I think it's due to applied suggestions :(

looks like this is resolved
